### PR TITLE
cache: Make the cache cloneable with RwLock

### PR DESCRIPTION
**Pull request by olix0r**
Caches are typically wrapped with buffers so that they can be shared
across connections. These buffers can be eliminated if the cache itself
becomes cloneable; and this is easily achieved with a
`parking_lot::RwLock`: the cache first obtains a read lock to clone the
inner service and a write lock is only obtained when the service is not
in the cache.

It turns out that the buffers have an important side effect, though: the
buffers hide the inner service type, which drastically reduces the
amount of memory needed to compile the linkerd2-proxy crate. To
replicate this, this change introduces a `stack::BoxNewService` which
can be used to hide the type signature of a `NewService`.

Benchmarks indicate that this improves p50 latency by 2ms for HTTP/2
traffic at 100 concurrenct requests.

### CODE REVIEW

File: linkerd/app/core/src/svc.rs [hunk 1]
```diff
@@ -284,6 +284,14 @@ impl<S> Stack<S> {
         self.push(http::boxed::response::Layer::new())
     }
 
+    pub fn box_new<T>(self) -> Stack<stack::BoxNewService<T, S::Service>>
```
Comment (hawkw) [inline linkerd/app/core/src/svc.rs:287]: nit/TIOLI: i might prefer to call this `box_new_service`...though it seems like we're not super consistent with naming the methods for new_service layers....

**Review by hawkw — APPROVED** [review]
this looks good to me, especially given that there's a performance improvement!

depending on the read/write ratio of the cache, i kinda wonder if it's possible to optimize this further for cases where the proxy has a larger number of worker threads with a sharded rwlock, but since this already makes everything faster it's probably not worth spending time on right now :)

File: Cargo.lock [hunk 2]
```diff
@@ -1773,8 +1814,19 @@ version = "0.10.2"
...
+[[package]]
+name = "parking_lot"
+version = "0.11.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "a4893845fa2ca272e647da5d0e46660a314ead9c2fdd9a883aabc32e481a8733"
+dependencies = [
+ "instant",
+ "lock_api 0.4.1",
+ "parking_lot_core 0.8.0",
 ]
```
[the remark this answers is not in the document]
Comment (olix0r) [reply Cargo.lock:1830]:
```
│   │   ├── tokio v0.2.22
│   │   │   ├── ...
│   │   │   ├── parking_lot v0.10.2
```

We could downgrade our dependency but I don't think it matters, practically.

[the remark this answers is not in the document]
Comment (hawkw) [reply Cargo.lock:1830]: :+1: not a big deal

**Pull request merged**